# Analysis of CoronaVirus Data

Starting off, lets import pandas.

Then I will upload the file, immediately changing NaN values to 0, changing the data type to integer, and sort by 'Total Cases' from the highest.

After than, I'll turn the df into a dictionary:

In [15]:
import pandas as pd


df = pd.read_csv('C:/Users/phpunsal/Documents/Data_Analytics/Corona_Virus_Data/data/corona.csv').fillna(0).astype(dtype=int, errors='ignore').sort_values(by='Total Cases', ascending=False)


latest = df.to_dict('list')

`latest` is now a dictionary, with each key as a column heading and each value is a list containing the values in the given column from each row:


In [16]:
latest

{'Country': ['China',
  'Diamond Princess',
  'Singapore',
  'Japan',
  'Hong Kong',
  'S. Korea',
  'Thailand',
  'USA',
  'Taiwan',
  'Malaysia',
  'Vietnam',
  'Germany',
  'Australia',
  'France',
  'Macao',
  'U.K.',
  'U.A.E.',
  'Canada',
  'Philippines',
  'Italy',
  'India',
  'Russia',
  'Spain',
  'Nepal',
  'Belgium',
  'Sri Lanka',
  'Finland',
  'Egypt',
  'Cambodia',
  'Sweden'],
 'Total Cases': [74187,
  621,
  81,
  80,
  63,
  51,
  35,
  29,
  23,
  22,
  16,
  16,
  15,
  12,
  10,
  9,
  9,
  8,
  3,
  3,
  3,
  2,
  2,
  1,
  1,
  1,
  1,
  1,
  1,
  1],
 'New Cases': [1751,
  79,
  0,
  6,
  1,
  20,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0],
 'Total Deaths': [2006,
  0,
  0,
  1,
  2,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0],
 'New Deaths': [138,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  

This shows the dictionary keys:

In [17]:
latest.keys()

dict_keys(['Country', 'Total Cases', 'New Cases', 'Total Deaths', 'New Deaths', 'Recovered', 'Serious'])

This accesses the column and cell values, in this case the `Country` column:

In [18]:
print(latest['Country'])

['China', 'Diamond Princess', 'Singapore', 'Japan', 'Hong Kong', 'S. Korea', 'Thailand', 'USA', 'Taiwan', 'Malaysia', 'Vietnam', 'Germany', 'Australia', 'France', 'Macao', 'U.K.', 'U.A.E.', 'Canada', 'Philippines', 'Italy', 'India', 'Russia', 'Spain', 'Nepal', 'Belgium', 'Sri Lanka', 'Finland', 'Egypt', 'Cambodia', 'Sweden']


This shows the elements at a given position in the same row:

In [19]:
print(latest['Country'][0])
print(latest['Total Cases'][0])
print(latest['Total Deaths'][0])
print(latest['Recovered'][0])

China
74187
2006
14796


In [20]:
print(latest['Country'][3])
print(latest['Total Cases'][3])
print(latest['Total Deaths'][3])
print(latest['Recovered'][3])

Japan
80
1
20


Here's a function to return the worldwide number of reported cases from the dictionary:

In [21]:
def case_count(data):  
    total = sum(data['Total Cases'])
    return total

In [22]:
case_count(latest)

75307

Another function can be created to return the number of countries that have reported cases:

In [23]:
def country_count(data):
    countries_count = len(data['Country'])
    return countries_count

In [24]:
country_count(latest)

30

Here, another function can be created to return the average number of cases over all listed countries, using the previous functions:

In [25]:
def average_cases(data):
    average = case_count(data)/country_count(data)
    return average

In [26]:
average_cases(latest)

2510.233333333333

Here, a further function can be created to return the number of countries where the 'Total cases' equals 1. This function involves a 'for' loop to iterate through the values, and an 'if' condition within the loop to check for cases that equal 1. There is then a counter to track the number of countries that match the desired criteria:

In [27]:
def single_case_country_count(data):
    count = 0
    
    for cases_count in data['Total Cases']:
        if cases_count == 1:
            count = count + 1
            
    return (count)

In [28]:
single_case_country_count(latest)

7

Next, a function is created to return a list of the countries where the number of cases is equal to 1.

This could help identify those nations where an outbreak may just be starting. In that case, a local or national lockdown may be recommended:

In [29]:
def single_case_countries(data):
    countries = []
    for country, cases in zip(data["Country"], data["Total Cases"]):
        if cases == 1:
            countries.append(country)
    return countries

In [30]:
single_case_countries(latest)

['Nepal', 'Belgium', 'Sri Lanka', 'Finland', 'Egypt', 'Cambodia', 'Sweden']

The next function returns a list of countries in which there are still active cases (where Total cases minus Total Deaths exceeds Recovered).

This can help identify those countries that may need categorising as High Risk:

In [31]:
def active_countries(data):
    countries = []
    for i, country in enumerate(data['Country']):
        if data['Total Cases'][i] - data['Total Deaths'][i] > data['Recovered'][i]:
            countries.append(country)
    return countries

In [32]:
active_countries(latest)

['China',
 'Diamond Princess',
 'Singapore',
 'Japan',
 'Hong Kong',
 'S. Korea',
 'Thailand',
 'USA',
 'Taiwan',
 'Malaysia',
 'Vietnam',
 'Germany',
 'Australia',
 'France',
 'Macao',
 'U.K.',
 'U.A.E.',
 'Canada',
 'Italy',
 'Egypt',
 'Sweden']

The next function returns a list of countries where there are no longer any active cases (where Total Cases minus Total Deaths equals Recovered).

This can help identify those countries where the virus appears to have disappeared from the population:

In [33]:
def cleared_countries(data):
    countries = []
    for i, country in enumerate(data['Country']):
        if data['Total Cases'][i] - data['Total Deaths'][i] == data['Recovered'][i]:
            countries.append(country)
    return countries

In [34]:
cleared_countries(latest)

['Philippines',
 'India',
 'Russia',
 'Spain',
 'Nepal',
 'Belgium',
 'Sri Lanka',
 'Finland',
 'Cambodia']

Next, lets upload some more Covid data, and specifically create a df for only 1st July 2020, from that data:

In [35]:
data = pd.read_csv('C:/Users/phpunsal/Documents/Data_Analytics/Corona_Virus_Data/world_data/owid-covid-data.csv')
data['date'] = pd.to_datetime(data['date'])
df = data[data['date'] == '2020-07-01']

df.head()

iso_code continent     location       date  total_cases  new_cases  \
173      AFG      Asia  Afghanistan 2020-07-01      31517.0      279.0   
300      ALB    Europe      Albania 2020-07-01       2535.0       69.0   
491      DZA    Africa      Algeria 2020-07-01      13907.0      336.0   
613      AND    Europe      Andorra 2020-07-01        855.0        0.0   
727      AGO    Africa       Angola 2020-07-01        284.0        8.0   

     total_deaths  new_deaths  total_cases_per_million  new_cases_per_million  \
173         746.0        13.0                  809.616                  7.167   
300          62.0         4.0                  880.881                 23.977   
491         912.0         7.0                  317.142                  7.662   
613          52.0         0.0                11065.812                  0.000   
727          13.0         2.0                    8.641                  0.243   

     ...  aged_70_older  gdp_per_capita  extreme_poverty  cvd_death_rate  \
173  ...          1.337        1803.987              NaN         597.029   
300  ...          8.643       11803.431              1.1         304.195   
491  ...          3.857       13913.839              0.5         278.364   
613  ...            NaN             NaN              NaN         109.135   
727  ...          1.362        5819.495              NaN         276.045   

     diabetes_prevalence  female_smokers  male_smokers  \
173                 9.59             NaN           NaN   
300                10.08             7.1          51.2   
491                 6.73             0.7          30.4   
613                 7.97            29.0          37.8   
727                 3.94             NaN           NaN   

     handwashing_facilities hospital_beds_per_thousand  life_expectancy  
173                  37.746                       0.50            64.83  
300                     NaN                       2.89            78.57  
491                  83.741                       1.90            76.88  
613                     NaN                        NaN            83.73  
727                  26.664                        NaN            61.15  

[5 rows x 34 columns]

In [36]:
df.tail()

iso_code continent        location       date  total_cases  new_cases  \
29411       ESH    Africa  Western Sahara 2020-07-01        380.0      172.0   
29506       YEM      Asia           Yemen 2020-07-01       1158.0       30.0   
29623       ZMB    Africa          Zambia 2020-07-01       1594.0       26.0   
29738       ZWE    Africa        Zimbabwe 2020-07-01        591.0       17.0   
29934  OWID_WRL       NaN           World 2020-07-01   10465987.0   192563.0   

       total_deaths  new_deaths  total_cases_per_million  \
29411           1.0         0.0                  636.164   
29506         312.0         8.0                   38.825   
29623          24.0         2.0                   86.706   
29738           7.0         0.0                   39.763   
29934      511041.0      5732.0                 1342.689   

       new_cases_per_million  ...  aged_70_older  gdp_per_capita  \
29411                287.948  ...          1.380             NaN   
29506                  1.006  ...          1.583        1479.147   
29623                  1.414  ...          1.542        3689.251   
29738                  1.144  ...          1.882        1899.775   
29934                 24.704  ...          5.355       15469.207   

       extreme_poverty  cvd_death_rate  diabetes_prevalence  female_smokers  \
29411              NaN             NaN                  NaN             NaN   
29506             18.8         495.003                 5.35           7.600   
29623             57.5         234.499                 3.94           3.100   
29738             21.4         307.846                 1.82           1.600   
29934             10.0         233.070                 8.51           6.434   

       male_smokers  handwashing_facilities hospital_beds_per_thousand  \
29411           NaN                     NaN                        NaN   
29506        29.200                  49.542                      0.700   
29623        24.700                  13.938                      2.000   
29738        30.700                  36.791                      1.700   
29934        34.635                  60.130                      2.705   

       life_expectancy  
29411            70.26  
29506            66.12  
29623            63.89  
29738            61.49  
29934            72.58  

[5 rows x 34 columns]

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 211 entries, 173 to 29934
Data columns (total 34 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   iso_code                         211 non-null    object        
 1   continent                        210 non-null    object        
 2   location                         211 non-null    object        
 3   date                             211 non-null    datetime64[ns]
 4   total_cases                      210 non-null    float64       
 5   new_cases                        210 non-null    float64       
 6   total_deaths                     210 non-null    float64       
 7   new_deaths                       210 non-null    float64       
 8   total_cases_per_million          210 non-null    float64       
 9   new_cases_per_million            210 non-null    float64       
 10  total_deaths_per_million         210 non-null    float64       

There seems to be a lot of missing data, especially when it comes to tests.

In [38]:
df.shape

(211, 34)

The dataframe consists of 34 columns and 211 rows.

The last row is for World, so this will have to be removed in a new df called Countries:

In [39]:
countries = df.copy()
countries = countries[countries['location']!='World']

Comparing the shapes of the DataFrames, it can be seen that the 1 row containing World was removed

In [40]:
print(df.shape, countries.shape)

(211, 34) (210, 34)


A new df based on the `countries` df is created, with only 3 columns and ordered by 'total_death_per_million':

In [41]:
cols = ['continent', 'location', 'total_deaths_per_million']

countries_dr = countries.copy()
countries_dr = countries_dr[cols].sort_values(by='total_deaths_per_million', ascending=False)

countries_dr


continent                          location  \
23306         Europe                        San Marino   
2917          Europe                           Belgium   
613           Europe                           Andorra   
28347         Europe                    United Kingdom   
25362         Europe                             Spain   
...              ...                               ...   
23111  North America  Saint Vincent and the Grenadines   
23926         Africa                        Seychelles   
15734         Africa                           Lesotho   
10808         Europe                         Gibraltar   
12195           Asia                         Hong Kong   

       total_deaths_per_million  
23306                  1237.551  
2917                    841.615  
613                     673.008  
28347                   644.168  
25362                   606.633  
...                         ...  
23111                     0.000  
23926                     0.000  
15734                     0.000  
10808                     0.000  
12195                       NaN  

[210 rows x 3 columns]

Now, using the 'countries' df, I can find the sum of the total tests for countries in the continent of Africa and convert the result into an integer:

In [42]:
africa_tests = countries[countries['continent']=='Africa']['total_tests'].sum().astype(int)
africa_tests

3445134

I can then find how many countries in Africa have no value recorded for total tests:

In [43]:
africa_missing_test_data = countries[countries['continent']=='Africa']['total_tests'].isna().sum()
africa_missing_test_data

45

Next, I can find how many countries have a higher value for total tests than the UK:

In [44]:
uk_tests = countries[countries['location']=='United Kingdom']['total_tests'].values[0]
countries_more_tests = len(countries[countries['total_tests']>uk_tests])
countries_more_tests

3

Then a new df is created, based on the `countries` df but with only 2 columns.

This will compare `hospital_beds_per_thousand` with `total_deaths_per_million`:

In [45]:
cols2 = ['hospital_beds_per_thousand', 'total_deaths_per_million']
beds_dr = countries.copy()
beds_dr = beds_dr[cols2].dropna()
beds_dr.head()

hospital_beds_per_thousand  total_deaths_per_million
173                         0.50                    19.163
300                         2.89                    21.544
491                         1.90                    20.798
952                         3.80                    30.635
1081                        5.00                    28.919

The `beds_dr` df is then used to identify the average `total_deaths_per_million` for entries where `hospital_beds_per_thousand` is greater than the mean.

The result is the ratio of Covid Deaths (per million) to Hospital Beds (per thousand), for countries with above average Hospital Beds per capita:

In [47]:
mean_beds = beds_dr['hospital_beds_per_thousand'].mean()

dr_high_bed_ratio = beds_dr[beds_dr['hospital_beds_per_thousand']>mean_beds]['total_deaths_per_million'].mean()
dr_high_bed_ratio

98.18423728813559

This shows for nations with an above average concentration of hospital beds, there is actually around 98 Covid related deaths (per million people) for every 1 hospital bed (per thousand people).

Lets now look at nations with a lower than average concentration of hospital beds:

In [44]:
dr_low_bed_ratio = beds_dr[beds_dr['hospital_beds_per_thousand']<mean_beds]['total_deaths_per_million'].mean()
dr_low_bed_ratio

56.294057142857135

Surprisingly, this shows that nations with a lower than average concentration of hospital beds have a much lower ratio of Covid related deaths (per million people), at around 56 deaths per Hospital bed (per thousand people).

This suggests that countries with less hospital beds have less Covid Deaths, although there may be other, economical reasons for this. Wealthier nations may spread the disease more rapidly. Perhaps hospitals were not able to prevent deaths so well?

Next, lets create a new df containing only rows of countries with zero new cases:

In [48]:
no_new_cases = countries[countries['new_cases'] == 0]
no_new_cases.head()

iso_code      continent             location       date  total_cases  \
613       AND         Europe              Andorra 2020-07-01        855.0   
836       AIA  North America             Anguilla 2020-07-01          3.0   
952       ATG  North America  Antigua and Barbuda 2020-07-01         66.0   
1381      ABW  North America                Aruba 2020-07-01        103.0   
2080      BHS  North America              Bahamas 2020-07-01        104.0   

      new_cases  total_deaths  new_deaths  total_cases_per_million  \
613         0.0          52.0         0.0                11065.812   
836         0.0           0.0         0.0                  199.973   
952         0.0           3.0         0.0                  673.965   
1381        0.0           3.0         0.0                  964.727   
2080        0.0          11.0         0.0                  264.464   

      new_cases_per_million  ...  aged_70_older  gdp_per_capita  \
613                     0.0  ...            NaN             NaN   
836                     0.0  ...            NaN             NaN   
952                     0.0  ...          4.631       21490.943   
1381                    0.0  ...          7.452       35973.781   
2080                    0.0  ...          5.200       27717.847   

      extreme_poverty  cvd_death_rate  diabetes_prevalence  female_smokers  \
613               NaN         109.135                 7.97            29.0   
836               NaN             NaN                  NaN             NaN   
952               NaN         191.511                13.17             NaN   
1381              NaN             NaN                11.62             NaN   
2080              NaN         235.954                13.17             3.1   

      male_smokers  handwashing_facilities hospital_beds_per_thousand  \
613           37.8                     NaN                        NaN   
836            NaN                     NaN                        NaN   
952            NaN                     NaN                        3.8   
1381           NaN                     NaN                        NaN   
2080          20.4                     NaN                        2.9   

      life_expectancy  
613             83.73  
836             81.88  
952             77.02  
1381            76.29  
2080            73.92  

[5 rows x 34 columns]

Then, the country with the highest number of total cases, but with no new cases, can be identified.

This can help identify those who have had success in stopping their outbreak:

In [49]:
highest_no_new = no_new_cases.loc[no_new_cases['total_cases'] == no_new_cases['total_cases'].max()]['location'].item()
highest_no_new

'Cameroon'

From the `countries` df, I can find the sum of the populations of all countries which have had zero total deaths, in millions:

In [50]:
sum_populations_no_deaths = (round((countries[countries['total_deaths']==0]['population'].sum())/1000000))
sum_populations_no_deaths

192

A function can be created, taking a dataframe, a location and a metric, and returning the location and the metric's value for that location.

This is helpful in pulling a specific bit of information from a vast dataset:

In [55]:
def country_metric(df, location, metric):
    
    return df[df['location'] == location].iloc[0][metric]

In [56]:
vietnam_older_70 = country_metric(countries, 'Vietnam', 'aged_70_older')
vietnam_older_70

4.718

Another function, that returns the mean of a metric for a group of countries.

Great for comparing a particular country against a group of countries:

In [57]:
def countries_average(df, countries, metric):

    df_countries = df['location'].isin(countries)
   
    
    meany = df[df_countries][metric].mean()
        
    return meany

Lets apply this function to the G7 Nations:

In [58]:
g7 = ['United States', 'Italy', 'Canada', 'Japan', 'United Kingdom', 'Germany', 'France']
g7_avg_life_expectancy = countries_average(df, g7, 'life_expectancy')
g7_avg_life_expectancy

82.10571428571428

Next, the country with the lowest life expectancy can be identified and compared against the G7 average:

In [61]:
lowest = countries[countries['life_expectancy'] == countries['life_expectancy'].min()].iloc[0]
country = lowest['location']
life_exp = lowest['life_expectancy']
diff = g7_avg_life_expectancy - life_exp
headline = f'{country} has a life expectancy of {round(diff,1)} years lower than the G7 average.'
print(headline)

Central African Republic has a life expectancy of 28.8 years lower than the G7 average.
